### PR TITLE
[TASK] avoid hash timing attacks

### PR DESCRIPTION
**Pull request by kaystrobach**

### CODE REVIEW

File: Classes/Hooks/Form.php [hunk 1]
```diff
@@ -100,7 +100,7 @@ public function afterSubmit(FormRuntime $runtime, $element, $value, $requestArgu
     protected function crFieldHasBeenVerified(FormRuntime $runtime): bool
     {
         return $runtime->getFormState() &&
-            $runtime->getFormState()->getFormValue(self::FIELD_ID) === sha1($this->getHmacSalt($runtime));
+            hash_equals($runtime->getFormState()->getFormValue(self::FIELD_ID), sha1($this->getHmacSalt($runtime)));
```
Comment (kaystrobach) [inline Classes/Hooks/Form.php:103]: we need to verify, that the first and second value of the comparision are not null.

**Comment by derhansen** [comment]
Can you explain the improvement this patch archives? I see no attack vector for a hash timing attack here, since submitted form values are not considered for the string comparison in `crFieldHasBeenVerified()`. The value in `$runtime->getFormState()->getFormValue(self::FIELD_ID)` is written with an internally known hash, when the CR response calculation was successful. So it is IMO safe to use strict string comparison here.

**Comment by kaystrobach** [comment]
> Can you explain the improvement this patch archives? I see no attack vector for a hash timing attack here, since submitted form values are not considered for the string comparison in `crFieldHasBeenVerified()`. The value in `$runtime->getFormState()->getFormValue(self::FIELD_ID)` is written with an internally known hash, when the CR response calculation was successful. So it is IMO safe to use strict string comparison here.

```php
sha1($this->getHmacSalt($runtime)
```

There is a hash calculation, which provides one of the values. That was the reason, why i stumbled and suggested to use the hash comparision.

**Comment by derhansen** [comment]
AFAICS `sha1($this->getHmacSalt($runtime))` does not include any user input. Can you point out, where this is evaluated?

**Comment by kaystrobach** [comment]
Form.php line 103. See above code line, which is highlighted in me comment

**Comment by derhansen** [comment]
Sorry, I still do not get the point. The value in `$runtime->getFormState()->getFormValue(self::FIELD_ID)` does not include any input that can be controlled by the user submitting the form. The value is set to an application internally calculated hash after successful validation of the CR field (see https://github.com/derhansen/form_crshield/blob/main/Classes/Hooks/Form.php#L94). Therefore, AFAICS the user submitting the form has never control of any input used for the generated hashes and can not perform any timing attack, which requires data-dependent timing information.

**Comment by kaystrobach** [comment]
Ok, then it should be fine. - sry for bothering you. Thank you for your time.

**Pull request closed without merging**